### PR TITLE
Separate Persistence Strategy and Serialization Strategy

### PR DESCRIPTION
**Pull request by schrockn**
This separates the notion of marshalling policy and serialization strategy. (Very open to feedback on names).

Serialization strategy are things like "pickling" or "json" or "parquet". File formats essentially.

Marshalling policy (could be called persistence) is more around concerns like "local file system" or "s3 bucket".

The idea is that types determine their own serialization strategy based on their characteristics, where as marshalling policy is more about where to stash values on a per-pipeline run basis.

### CODE REVIEW

**Comment by schrockn** [comment]
I’m going to wait to merge this until my other big merges land

Sent via Superhuman iOS ( https://sprh.mn/?vip=schrockn@elementl.com )

**Comment by schrockn** [comment]
So please don’t merge this. I’ll do it

Sent via Superhuman iOS ( https://sprh.mn/?vip=schrockn@elementl.com )

**Comment by mgasner** [comment]
I did a little bit of looking around and while these terms are very confused, there does seem to be rough agreement that "serialization" and "marshalling" are similar ideas having to do with turning in-memory objects into representations that can be written out as bytestreams.

Sometimes these are synonyms, sometimes marshalling is understood to be more type-aware, remote environment-aware, or codebase-aware -- so that unmarshalling an object involves an additional step of loading up class definitions and instantiating an object from them. (https://en.wikipedia.org/wiki/Marshalling_(computer_science)#Comparison_with_serialization)

I think marshalling in this sense is something we want to avoid evoking. So I would prefer that we use the term "serialization" to refer to the process of changing in-memory objects into formatted bytestreams, and use the term "persistence" to refer to how and where we write, store, or output the serialized objects.

File: python_modules/dagster/dagster/core/types/marshal.py [hunk 1]
```diff
@@ -3,23 +3,63 @@
...
+
+class PickleSerializationStrategy(SerializationStrategy):
+    def serialize_value(self, value, write_file_obj):
+        pickle.dump(value, write_file_obj)
+
+    def deserialize_value(self, read_file_obj):
+        return pickle.load(read_file_obj)
+
+
+@six.add_metaclass(ABCMeta)
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/types/marshal.py]: very fun

**Comment by schrockn** [comment]
I can switch to persistence np

Sent via Superhuman ( https://sprh.mn/?vip=schrockn@elementl.com )

**Pull request merged**